eclair-core/src/main/scala/fr/acinq/eclair/router/RouteCalculation.scala

```diff
@@ -324,15 +324,24 @@ object RouteCalculation {
                                          routeParams: RouteParams,
                                          currentBlockHeight: Long): Either[RouterException, Seq[Route]] = {
     // We use Yen's k-shortest paths to find many paths for chunks of the total amount.
-    val numRoutes = {
-      val directChannelsCount = g.getEdgesBetween(localNodeId, targetNodeId).length
-      routeParams.mpp.maxParts.max(directChannelsCount) // if we have direct channels to the target, we can use them all
+    // When the recipient is a direct peer, we have complete visibility on our local channels so we can use more accurate MPP parameters.
+    val routeParams1 = {
+      case class DirectChannel(balance: MilliSatoshi, isEmpty: Boolean)
+      val directChannels = g.getEdgesBetween(localNodeId, targetNodeId).map(e => {
+        val balance = e.balance_opt.getOrElse(e.capacity.toMilliSatoshi)
+        val isEmpty = balance < e.update.htlcMinimumMsat
+        DirectChannel(balance, isEmpty)
+      })
+      // If we have direct channels to the target, we can use them all.
+      val numRoutes = routeParams.mpp.maxParts.max(directChannels.length)
+      // If we have direct channels to the target, we can use them all, even if they have only a small balance left.
+      val routeAmount = (amount +: routeParams.mpp.minPartAmount +: directChannels.filter(!_.isEmpty).map(_.balance)).min
```

Comment:
I don't get that line:
```scala
val routeAmount = (amount +: routeParams.mpp.minPartAmount +: directChannels.filter(!_.isEmpty).map(_.balance)).min
```

Reply:
We want to use the lowest common denominator between:

- the total amount we want to send
- the configured minimum part amount
- the lowest balance of our our usable channels when the recipient is a direct peer

That's why we put these all in a `Seq` and take the min.

Reply:
The path-finding algorithm will find a bunch of non-overlapping routes that can carry that amount, then the `split` method will "fill" each of these routes with a bigger amount to send depending on the route's capacity.

Reply:
I think the naming is what got me confused:

```suggestion
      val minPartAmount = (amount +: routeParams.mpp.minPartAmount +: directChannels.filter(!_.isEmpty).map(_.balance)).min
```

Reply:
Right, that makes sense, fixed in https://github.com/ACINQ/eclair/pull/1599/commits/3ee956d451af173399db4d92878aa3acdf960342